File: app/src/main/java/com/krystianwsul/checkme/gui/tasks/MonthlyWeekScheduleEntry.java
package com.krystianwsul.checkme.gui.tasks;

import android.content.Context;
import android.os.Parcel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.loaders.CreateTaskLoader;
import com.krystianwsul.checkme.utils.CustomTimeKey;
import com.krystianwsul.checkme.utils.ScheduleType;
import com.krystianwsul.checkme.utils.Utils;
import com.krystianwsul.checkme.utils.time.Date;
import com.krystianwsul.checkme.utils.time.DayOfWeek;
import com.krystianwsul.checkme.utils.time.TimePair;
import com.krystianwsul.checkme.utils.time.TimePairPersist;

import junit.framework.Assert;

import java.util.Collections;
import java.util.Map;

class MonthlyWeekScheduleEntry extends ScheduleEntry {
    private final int mMonthWeekNumber;

    @NonNull
    private final DayOfWeek mMonthWeekDay;

    private final boolean mBeginningOfMonth;

    @NonNull
    private final TimePair mTimePair;

    MonthlyWeekScheduleEntry(@NonNull CreateTaskLoader.ScheduleData.MonthlyWeekScheduleData monthlyWeekScheduleData) {
        mMonthWeekNumber = monthlyWeekScheduleData.getDayOfMonth();
        mMonthWeekDay = monthlyWeekScheduleData.getDayOfWeek();
        mBeginningOfMonth = monthlyWeekScheduleData.getBeginningOfMonth();
        mTimePair = monthlyWeekScheduleData.getTimePair().copy();
    }

    private MonthlyWeekScheduleEntry(int monthWeekNumber, @NonNull DayOfWeek monthWeekDay, boolean beginningOfMonth, @NonNull TimePair timePair, @Nullable String error) {
        super(error);

        mMonthWeekNumber = monthWeekNumber;
        mMonthWeekDay = monthWeekDay;
        mBeginningOfMonth = beginningOfMonth;
        mTimePair = timePair;
    }

    MonthlyWeekScheduleEntry(@NonNull ScheduleDialogFragment.ScheduleDialogData scheduleDialogData) {
        Assert.assertTrue(scheduleDialogData.getMScheduleType() == ScheduleType.MONTHLY_WEEK);
        Assert.assertTrue(!scheduleDialogData.getMMonthlyDay());

        mMonthWeekNumber = scheduleDialogData.getMMonthWeekNumber();
        mMonthWeekDay = scheduleDialogData.getMMonthWeekDay();
        mBeginningOfMonth = scheduleDialogData.getMBeginningOfMonth();
        mTimePair = scheduleDialogData.getMTimePairPersist().getTimePair();
    }

    @NonNull
    @Override
    public String getText(@NonNull Map<CustomTimeKey, CreateTaskLoader.CustomTimeData> customTimeDatas, @NonNull Context context) {
        String day = Utils.ordinal(mMonthWeekNumber) + " " + mMonthWeekDay + " " + context.getString(R.string.monthDayStart) + " " + context.getResources().getStringArray(R.array.month)[mBeginningOfMonth ? 0 : 1] + " " + context.getString(R.string.monthDayEnd);

        if (mTimePair.mCustomTimeKey != null) {
            Assert.assertTrue(mTimePair.mHourMinute == null);

            CreateTaskLoader.CustomTimeData customTimeData = customTimeDatas.get(mTimePair.mCustomTimeKey);
            Assert.assertTrue(customTimeData != null);

            return day + ", " + customTimeData.getName();
        } else {
            Assert.assertTrue(mTimePair.mHourMinute != null);

            return day + ", " + mTimePair.mHourMinute.toString();
        }
    }

    @NonNull
    @Override
    public CreateTaskLoader.ScheduleData getScheduleData() {
        return new CreateTaskLoader.ScheduleData.MonthlyWeekScheduleData(mMonthWeekNumber, mMonthWeekDay, mBeginningOfMonth, mTimePair);
    }

    @NonNull
    @Override
    public ScheduleDialogFragment.ScheduleDialogData getScheduleDialogData(@NonNull Date today, @Nullable CreateTaskActivity.ScheduleHint scheduleHint) {
        Date date = (scheduleHint != null ? scheduleHint.getMDate() : today);

        date = Utils.getDateInMonth(date.getYear(), date.getMonth(), mMonthWeekNumber, mMonthWeekDay, mBeginningOfMonth);

        return new ScheduleDialogFragment.ScheduleDialogData(date, Collections.singleton(mMonthWeekDay), false, date.getDay(), mMonthWeekNumber, mMonthWeekDay, mBeginningOfMonth, new TimePairPersist(mTimePair), ScheduleType.MONTHLY_WEEK);
    }

    @NonNull
    @Override
    public ScheduleType getScheduleType() {
        return ScheduleType.MONTHLY_WEEK;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(mMonthWeekNumber);
        parcel.writeSerializable(mMonthWeekDay);
        parcel.writeInt(mBeginningOfMonth ? 1 : 0);
        parcel.writeParcelable(mTimePair, 0);
        parcel.writeString(getError());
    }

    @SuppressWarnings("unused")
    public static final Creator<MonthlyWeekScheduleEntry> CREATOR = new Creator<MonthlyWeekScheduleEntry>() {
        @Override
        public MonthlyWeekScheduleEntry createFromParcel(Parcel in) {
            int monthWeekNumber = in.readInt();

            DayOfWeek monthWeekDay = (DayOfWeek) in.readSerializable();
            Assert.assertTrue(monthWeekDay != null);

            boolean beginningOfMonth = (in.readInt() == 1);

            TimePair timePair = in.readParcelable(TimePair.class.getClassLoader());
            Assert.assertTrue(timePair != null);

            String error = in.readString();

            return new MonthlyWeekScheduleEntry(monthWeekNumber, monthWeekDay, beginningOfMonth, timePair, error);
        }

        @Override
        public MonthlyWeekScheduleEntry[] newArray(int size) {
            return new MonthlyWeekScheduleEntry[size];
        }
    };
}

File: app/src/main/java/com/krystianwsul/checkme/gui/tasks/MonthlyWeekScheduleEntry.kt
package com.krystianwsul.checkme.gui.tasks

import android.content.Context
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.utils.CustomTimeKey
import com.krystianwsul.checkme.utils.ScheduleType
import com.krystianwsul.checkme.utils.Utils
import com.krystianwsul.checkme.utils.time.Date
import com.krystianwsul.checkme.utils.time.DayOfWeek
import com.krystianwsul.checkme.utils.time.TimePair
import com.krystianwsul.checkme.utils.time.TimePairPersist
import com.krystianwsul.checkme.viewmodels.CreateTaskViewModel


class MonthlyWeekScheduleEntry(monthlyWeek: CreateTaskViewModel.ScheduleData.MonthlyWeek) : ScheduleEntry() {

    private val monthWeekNumber: Int

    private val monthWeekDay: DayOfWeek

    private val beginningOfMonth: Boolean

    private val timePair: TimePair

    override val scheduleData: CreateTaskViewModel.ScheduleData
        get() = CreateTaskViewModel.ScheduleData.MonthlyWeek(monthWeekNumber, monthWeekDay, beginningOfMonth, timePair)

    override val scheduleType = ScheduleType.MONTHLY_WEEK

    init {
        monthWeekNumber = monthlyWeek.dayOfMonth
        monthWeekDay = monthlyWeek.dayOfWeek
        beginningOfMonth = monthlyWeek.beginningOfMonth
        timePair = monthlyWeek.timePair.copy()
    }

    override fun getText(customTimeDatas: Map<CustomTimeKey<*>, CreateTaskViewModel.CustomTimeData>, context: Context): String {
        val day = Utils.ordinal(monthWeekNumber) + " " + monthWeekDay + " " + context.getString(R.string.monthDayStart) + " " + context.resources.getStringArray(R.array.month)[if (beginningOfMonth) 0 else 1] + " " + context.getString(R.string.monthDayEnd)

        return "$day, " + if (timePair.customTimeKey != null) {
            check(timePair.hourMinute == null)

            val customTimeData = customTimeDatas.getValue(timePair.customTimeKey)

            customTimeData.name
        } else {
            timePair.hourMinute!!.toString()
        }
    }

    override fun getScheduleDialogData(today: Date, scheduleHint: CreateTaskActivity.Hint.Schedule?): ScheduleDialogFragment.ScheduleDialogData {
        var date = scheduleHint?.date ?: today

        date = Utils.getDateInMonth(date.year, date.month, monthWeekNumber, monthWeekDay, beginningOfMonth)

        return ScheduleDialogFragment.ScheduleDialogData(date, mutableSetOf(monthWeekDay), false, date.day, monthWeekNumber, monthWeekDay, beginningOfMonth, TimePairPersist(timePair), ScheduleType.MONTHLY_WEEK)
    }
}
